Added backport.yml to automate backport process

siddharthasonker95 opened the pull request:
Objective: Automate the backport process.

Description: 

Add the label like `backport <branch_name>` to the pull request that needs to be backported to a specific branch. Once the pull request is successfully merged, the backport workflow will create a pull request with the required changes to be merged into the selected branch.

Changes:
- Added `backport.yml` and `/scripts/backport/main.sh` to automate `backport` process. 
<!--

Describe in detail the changes you are proposing, and the rationale.

See the contributing guide:

https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md

-->

<!--

Link all GitHub issues fixed by this PR, and add references to prior related PRs.
Make sure to first open an issue, get community approval and only then create Pull Request to resolve it.
All Pull Requests must have an issue attached to them

-->

Resolves #1112 

## Target Release

<!--

In normal circumstances we only target changes at the upcoming minor
release, or as a patch to the current minor version. If you need to
port a security fix to an older release, highlight this here by listing
all targeted releases.

-->

1.8.0

Yantrio commented:
Please dont forget to sign off on the DCO @siddharthasonker95 :)

Yantrio commented:
I think it may be best to also implement the backporting from a comment, as per these docs: https://github.com/korthout/backport-action?tab=readme-ov-file#trigger-using-a-comment


also we could configure it to copy assignees/reviewers too. That would be helpful.

janosdebugs commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,31 @@
+    steps:
+      - uses: actions/checkout@v4
+      - name: Create backport pull requests
+        uses: korthout/backport-action@v2
```
Adding GitHub Actions like this is a security risk as they get GitHub access tokens. This action isn't well known and seems to be maintained by a single person. I would advise against using it in this fashion. In fact, I would strongly prefer not to introduce any more GitHub Actions as we should actually lock down the org to prevent unintended actions from being added.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+git config --global user.email "noreply@github.com"
+git config --global user.name "GitHub Actions"
+
+git config --global credential.helper "store --file=~/.git-credentials"
```
I don't believe you need to configure git credentials when on a github runner with the correct setup.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+echo "Successfully authenticated!"
+
+git checkout $GITHUB_EVENT_BRANCH
+echo "Checked out main"
```
This does not match the previous command

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+labels=$(jq --raw-output '.pull_request.labels[].name' "$GITHUB_EVENT_PATH")
+backport_label=""
+branch_name=""
+for label in "${labels[@]}"; do
```
Can github labels have spaces and does this handle that case?

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
```
Let's include set -e so that if a command exits with a non-zero status code, the script fails.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+echo "Label Name: $backport_label"
+
+# Checkout the version branch
+git fetch --all
```
Why --all and not just the branch name?

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+echo "Version branch: $branch_name"
+
+# Checkout new backport branch
+ISSUE_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
```
Let's do all of the issue parsing in the same location in the script for clarity.  Also, I'd prefer to do as much of this as possible in the github workflow file and not in a bash script.

cam72cam replied:
Also why is this uppercase when all the rest of the vars are lower case?

siddharthasonker95 replied:
I tested the commands on my terminal on personal repo and didn't notice the variable naming consistency.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+
+# Checkout new backport branch
+ISSUE_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
+newBranch="backport/$ISSUE_NUMBER"
```
Why is this camelCase when other vars are underscore_case?

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+echo "New backport branch: $newBranch"
+
+# Get pull request and repository info.
+pullRequestNumber=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
```
Same comment as above, we should do most of this in the workflow file.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+done
+
+# Count the number of commits
+commitCount=$(echo "$commitIDs" | wc -l)
```
This is not nessesary

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+ISSUE_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
+newBranch="backport/$ISSUE_NUMBER"
+git checkout -b $newBranch
+git push origin $newBranch
```
Why do we push this branch before making any changes?

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,90 @@
+git push origin $newBranch
+echo "Successfully pushed!"
+
+head_branch=$(jq -r '.pull_request.head.ref' "$GITHUB_EVENT_PATH")
```
Same comment on doing this in the workflow

cam72cam commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,33 @@
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        GITHUB_EVENT: ${{ toJson(github.event) }} # Pass the github.event object as a JSON string
+      run: |
+        cd $GITHUB_WORKSPACE  # Move to the root of the repository
```
This is not necessary

cam72cam replied:
@siddharthasonker95 why did you resolve this?

siddharthasonker95 replied:
I assumed it was related to this [comment](https://github.com/opentofu/opentofu/pull/1452/files#r1576121524) and have replied to it.

cam72cam commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,33 @@
+        GITHUB_EVENT: ${{ toJson(github.event) }} # Pass the github.event object as a JSON string
+      run: |
+        cd $GITHUB_WORKSPACE  # Move to the root of the repository
+        cd .github/scripts/backport
```
Why cd into the scripts directory when you can `./.github/scripts/backport/main.sh` instead?

siddharthasonker95 replied:
I tried using `./.github/scripts/backport/main.sh` first in my repo but somehow it was not able to locate the directory.

siddharthasonker95 replied:
updated.

cam72cam reviewed (dismissed):
Looks like a reasonable start, but I've added a number of comments.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,66 @@
+# Get backport label from all the labels of the PR
+backport_label=""
+branch_name=""
+for label in "${LABELS[@]}"; do
```
Question for @cube2222, do we want to support backporting to multiple versions?

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,66 @@
+
+# Store the commit SHAs in an array
+commit_ids_array=($commit_ids)
+for commit_sha in "${commit_ids_array[@]}"; do
```
Why is this using an array instead of simple iteration?

siddharthasonker95 replied:
done

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,66 @@
+# Store the commit SHAs in an array
+commit_ids_array=($commit_ids)
+for commit_sha in "${commit_ids_array[@]}"; do
+    git fetch origin $commit_sha
```
Is this git fetch needed?

siddharthasonker95 replied:
Yes, without `git fetch`, it was not able to detect the commit ID. That's why I added `git fetch` for every commit ID for safety.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,66 @@
+
+# Create the pull request.
+title="Backport PR #$ISSUE_NUMBER"
+body="This pull request backports changes from $head_branch to $branch_name."
```
head_branch is not defined.

siddharthasonker95 replied:
done.

cam72cam commented:
Another question for @cube2222: Do we want to cherry-pick individual commits or the merge commit?

cube2222 commented:
@cam72cam Since we squash, I don't believe there's a distinction here, no? That is, there's just a single commit for each change.

cam72cam commented:
I'm curious if it would make a merge conflict easier to spot with the merge commit vs all the branch commits.

siddharthasonker95 commented:
IMO, detecting merge conflicts from a merge commit will be more challenging compared to handling conflicts from individual commits one by one. Another workaround, if we choose to cherry-pick each commit for backporting, is to have the script fail when a conflict occurs during cherry-picking. Additionally, it will add a comment to the existing pull request with the commit ID that caused the conflict.
cc @cam72cam @cube2222

siddharthasonker95 commented:
I have addressed the PR comments and added the changes where, if the backport script fails to cherry-pick a commit due to a merge conflict, it will add a comment to the existing PR, specifying the commit ID that caused the conflict.
PTAL @cam72cam

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,78 @@
+set -e
+
+# Setting up Committer identity.
+git config --global user.email "noreply@github.com"
```
Let's remove the --global here just in case someone runs this locally.

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,78 @@
+fi
+
+# Checkout the version branch
+git fetch origin "$branch_name"
```
Why are these steps necessary?

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,78 @@
+echo "Version branch: $branch_name"
+
+# Checkout new backport branch
+new_branch="backport/$ISSUE_NUMBER"
```
We should include the label in the branch name.

siddharthasonker95 replied:
Sure, as the label will have format like `backport <branch_name>`, then we can name the backport branch like `backport/<branch_name>/<issue_number>`

cam72cam commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,78 @@
+# Create the pull request.
+title="Backport PR #$ISSUE_NUMBER"
+body="This pull request backports changes from $HEAD_BRANCH to $branch_name."
+url=$(curl -sS -X POST -H "Authorization: token $GITHUB_TOKEN" \
```
Why did you remove the usage of the `gh` command?  If you set GH_TOKEN correctly, it should work just fine.  I'd prefer that over constructing this inline.

siddharthasonker95 replied:
I have replaced the `gh` command changes to maintain consistency in the code, as all the other git actions have.

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,78 @@
+echo "Label Name: $backport_label"
+if [ -z "$backport_label" ]; then
+    echo "Backport label not found. Exiting."
+    exit 1
```
Should it return `exit 1` or `exit 0` when backport label is not present?
As this script runs whenever a PR is closed, and the backport label is present, then the rest of the script will run; otherwise, it will fail.
cc @janosdebugs @cam72cam

cam72cam replied:
exit 0

siddharthasonker95 commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,31 @@
+        echo "REPO=${{ github.event.repository.name }}" >> $GITHUB_ENV
+        echo "ISSUE_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+        echo "HEAD_BRANCH=${{ github.event.pull_request.head.ref }}" >> $GITHUB_ENV
+        echo "LABELS=$(jq -r '.pull_request.labels[].name' "$GITHUB_EVENT_PATH")" >> $GITHUB_ENV
```
I have separated setting the variable from the script execution step for better readability.

siddharthasonker95 commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,31 @@
+    name: Backport pull request
+    runs-on: ubuntu-latest
+    if: >
+      github.event.pull_request.merged && ${{contains(github.event.pull_request.labels.*.name, 'backport')}}
```
added readme file [here.](https://github.com/opentofu/opentofu/pull/1452/commits/23d19654d9b27e11be9850409fc14c0dd5dd99b4)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,79 @@
+        echo "Cherry-picking commit: ${commit_sha}"
+    else
+        # Add a failure comment to the pull request
+        echo "Error: Failed to cherry-pick commit ${commit_sha}"
```
fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,79 @@
+fi
+
+# Checkout the version branch
+echo "Fetching branch ${branch_name}..."
```
I have added the info and error logs in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,79 @@
+
+# Exit if backport label not found
+if [ -z "$backport_label" ]; then
+    echo "Warning: The backport label is not found on the Pull Request. Skipping the backport process"
```
done.

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,79 @@
+# Create the pull request.
+title="Backport PR #$ISSUE_NUMBER"
+body="This pull request backports changes from $HEAD_BRANCH to $branch_name."
+url=$(gh pr create --title "$title" --body "$body" --base "$branch_name" --head "$new_branch")
```
handled the error in the latest [commit.](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,79 @@
+done
+echo "Cherry-picking process completed successfully for pull request #${PR_NUMBER}."
+
+git push origin "$new_branch"
```
I have handled the error in every git commands in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
+
+set -e
```
fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+}
+
+# Setting up Committer identity.
+log_info "Setting up Committer identity..."
```
fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+        exit 1
+    fi
+
+    echo "Error: $1" >&2
```
fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+# Function to handle errors
+log_error() {
+    if [ -z "$1" ]; then
+        echo "Error: No error message provided." >&2
```
fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+# Function to output info logs
+log_info() {
+    if [ -z "$1" ]; then
+        echo "No message provided."
```
fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+        return
+    fi
+
+    echo "$1"
```
fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+fi
+
+if ! git config user.name "GitHub Actions"; then
+    log_error "Failed to set username"
```
fixed

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+    log_error "Authentication check for 'gh' failed."
+fi
+
+if [ -z "$OWNER" ] || [ -z "$REPO" ] || [ -z "$PR_NUMBER" ] || [ -z "$ISSUE_NUMBER" ] || [ -z "$HEAD_BRANCH" ]; then
```
fixed in the latest commit.

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+    log_error "Failed to create new backport branch ${new_branch}"
+fi
+
+if ! git push origin "$new_branch"; then
```
added.

siddharthasonker95 commented on .github/workflows/backport.yml:
```diff
@@ -0,0 +1,28 @@
+on:
+  pull_request:
+    types:
+      - closed
```
I have updated the backport.yml to trigger the backport script when a pull request is closed or labeled and pull request should fulfill following condition:
1. Pull request should be closed
2. Pull request should be merged
3. Pill request should have a label containing `backport` string.

siddharthasonker95 replied:
Below are the improvements that can be made to ensure the script's idempotence:

1. Validate whether the branch exists before creating it.
2. Check if the pull request already exists before attempting to create it.
3. Verify if a comment has been added to the pull request before attempting to add another.
4. Validate whether the branch already contains cherry-picked commits.

All validations are straightforward except for validating cherry-picked commits. If you have any suggestions on how to perform this validation, please let me know.

siddharthasonker95 replied:
I have implemented the suggested checks. Specifically, I added a check for branch existence; if the branch exists, it deletes the branch with all temporary changes, recreates it, and restarts the backport process.

I chose not to use `git rebase` because it is more error-prone and can potentially cause conflicts or unintended changes. Instead, I opted for branch deletion and recreation to ensure a clean and stable backport process.

siddharthasonker95 replied:
Additionally, I included a check to verify if a pull request exists between the target branch and the new backport branch. If a pull request exists, the backport process is skipped; otherwise, it continues. 
I also added a manual test case as suggested.
PTAL again.

siddharthasonker95 commented on .github/scripts/backport/main.sh:
```diff
@@ -0,0 +1,147 @@
+fi
+
+# Check if gh is logged in
+if ! gh auth status >/dev/null 2>&1; then
```
fixed in this [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)
